datafusion/physical-optimizer/src/optimizer.rs

```diff
@@ -131,6 +131,8 @@ impl PhysicalOptimizer {
             // replacing operators with fetching variants, or adding limits
             // past operators that support limit pushdown.
             Arc::new(LimitPushdown::new()),
+            // This FilterPushdown handles dynamic filters that may have references to the source ExecutionPlan
+            Arc::new(FilterPushdown::new_post_optimization()),
```

Comment:
this is so much nicer than adding it to `EnforceSorting` ❤️ 

But shouldn't this be the final pass? (maybe right before SanityCheckPlan?)

As I understand it, this filter pushdown pass has to be run after any pass that modifies the structure of the plan and `ProjectionPushdown` may actually do that 🤔 

It also think it would be good to add a comment here explaining that `FilterPushdown::new_post_optimization()` must be run after all passes that change the structure of the plan as it can generate pointers from one plan to another